graders/r/r_autograder/pltest.R

```diff
@@ -12,31 +18,78 @@ result <- tryCatch({
+        if (isFALSE(ttres) && inherits(ttres, "ttvd")) { 	# when false and a 'visual diff' result
+            ad <- attr(ttres, "diffplot")                   	# where the diff attribute is not NA
+            if (!is.na(ad) && grepl("^data:image/png", ad)) {   # and begins as base64 encoded image
+                D <- data.frame(label = "Difference Between Submitted and Expected Plot: Red Shows Difference",
+                                url = ad) 			# extract the encodes image (in 'ad')
+                images <- rbind(images,	D)			# and store in images
+            }
+        } else if (isFALSE(ttres) && inherits(ttres, "ttdo")) {
+            ## For false answer we collate call and diff output (from diffobj::diffPrint) below
+            if (debug) cat("[pltest] Flagging 'ttdo'\n")
+            ttdos[i, "output"] <- TRUE 				# pass an explicit flag
+        }
+        raw_test_results[[i]] <- ttres                  	# and update results
+    }
+    
+    test_results <- as.data.frame(raw_test_results)
+    test_results <- cbind(test_results, ttdos)
+    if (debug) {
+        cat("[pltest] showing test_results\n")
+        print(test_results)
+        cat("[pltest] done showing test_results\n\n")
+    }
+    
     ## Aggregate test results and process NAs as some question may have exited
-    test_results <- test_results[with(test_results, order(file, first)), ]
-    res <- cbind(test_results, question_details)
+    res <- merge(test_results, question_details, by = "file", all = TRUE)
     ## Correct answers get full points, other get nothing
     res$points <- ifelse( !is.na(res$result) & res$result==TRUE,  res$max_points, 0)
-    ## For false answer we collate call and diff output (from diffobj::diffPrint)
-    res$output <- ifelse( !is.na(res$result) & res$result==FALSE,
+    ## For false answer with ttdo support we collate call and diff output (from diffobj::diffPrint)
+    res$output <- ifelse( (!is.na(res$result) & res$result==FALSE) | res$output,
                          paste(res$call, res$diff, sep = "\n"), "")
+    ## We aggregate the score
     score <- base::sum(res$points) / base::sum(res$max_points) # total score
 
-    ## Columns needed by PL
+    if (debug) {
+        cat("[pltest] showing str(res)\n")
+        print(str(res))
+        cat("[pltest] done showing str(res)\n\n")
+    }
+
+    
+    ## Subset to columns needed by PL
     res <- res[, c("name", "max_points", "points", "output")]
 
-    ## output
-    list(tests = res, score = score, succeeded = TRUE)
+    ## And return components for JSON output
+    list(tests = res, score = score, images = images, succeeded = TRUE)
 },
 warning = function(w) list(tests = plr::message_to_test_result(w), score = 0, succeeded = FALSE),
 error = function(e) list(tests = plr::message_to_test_result(e), score = 0, succeeded = FALSE) )
 
 ## Record results as the required JSON object
 jsonlite::write_json(result, path = "results.json", auto_unbox = TRUE, force = TRUE)
+
+
```

Comment:
Nit, avoid multiple empty lines at the end of the file.

Reply:
Good catch, fixed in 0c271a216,